### ISSUE TITLE
Wrap Socket options

### ISSUE THREAD

**Issue by jcockayne** [azawawi/SwiftyZeroMQ#7]
It looks like pub/sub doesn't work at the moment without first calling `zmq_setsockopt` on the subscriber. This is because the default for .subscribe sockets is to ignore all messages. We should start looking at wrapping [zmq_setsockopt](http://api.zeromq.org/2-1:zmq-setsockopt) to get around this. 

Providing a generic function as we have in the `Context` is not going to be so simple because, depending on the option name, different data types can be passed. I would suggest we either provide a method for each option name, or a method for each data type - the former seems neater but requires more boilerplate.

**Comment by azawawi** [comment]
Nice suggestion. For example, `ZMQ_AFFINITY`  can be `socket.setAffinity`. I was going to start working on it but I am currently on vacation without any access to a Mac. Please feel free to submit a pull request.

**Comment by azawawi** [comment]
@jcockayne Any update on this one?

**Comment by azawawi** [comment]
@jcockayne  Please also note the following new socket options in ZeroMQ 4.2.0 (updated today)

```

    ZMQ_HANDSHAKE_IVL
    ZMQ_SOCKS_PROXY
    ZMQ_XPUB_NODROP
    ZMQ_XPUB_MANUAL
    ZMQ_XPUB_WELCOME_MSG
    ZMQ_STREAM_NOTIFY
    ZMQ_INVERT_MATCHING
    ZMQ_HEARTBEAT_IVL
    ZMQ_HEARTBEAT_TTL
    ZMQ_HEARTBEAT_TIMEOUT
    ZMQ_XPUB_VERBOSER
    ZMQ_CONNECT_TIMEOUT
    ZMQ_TCP_MAXRT
    ZMQ_THREAD_SAFE
    ZMQ_MULTICAST_MAXTPDU
    ZMQ_VMCI_BUFFER_SIZE
    ZMQ_VMCI_BUFFER_MIN_SIZE
    ZMQ_VMCI_BUFFER_MAX_SIZE
    ZMQ_VMCI_CONNECT_TIMEOUT
    ZMQ_USE_FD
```

**Comment by jcockayne** [comment]
Yes, sorry I am getting there! I did a large number of them this weekend (see my fork, these are predominantly those described in the old [ZMQ docs](http://api.zeromq.org/2-1:zmq-setsockopt)), but as you pointed out there are more that I have not yet wrapped. I'll try to find some time to do this over the next few days.

**Comment by azawawi** [comment]
@jcockayne I saw your fork earlier. Great work :+1:

**Comment by azawawi** [comment]
@jcockayne  What about I manually merge https://github.com/jcockayne/SwiftyZeroMQ/commit/4fa93d81d2c79ae6fc035590c138d236749cd532? 👍

**Comment by jcockayne** [comment]
Sure - or I can just submit a pull request if that's easier? Even though it's not yet completed a lot of the important ones are wrapped and the pattern is there!

**Comment by azawawi** [comment]
Go for it 👍

**Comment by erparker** [comment]
I'm very interested in the progress of this PR, thanks for giving this issue some attention! Any chance there will be a pub-sub example added to this repo once the PR is merged?

**Comment by azawawi** [comment]
@jcockayne I will be integrating the changes in your PR with master.
@erparker thanks for the interest, a pub-sub example will be added soon 👍

**Comment by azawawi** [comment]
@erparker a pub-sub example is released with 1.0.24
@jcockayne Thanks for your contribution.

Have fun.
